File: src/main/java/dragonsaga/ui/RaceChangeStatForm.java
package dragonsaga.ui;

import dragonsaga.ui.components.FormHorizontalFloatScroll;
import dragonsaga.ui.models.StatElementData;
import necesse.engine.localization.message.LocalMessage;
import necesse.gfx.forms.Form;
import necesse.gfx.forms.components.*;
import necesse.gfx.forms.components.localComponents.FormLocalLabel;
import necesse.gfx.gameFont.FontOptions;

public class RaceChangeStatForm extends Form {

    private FormLocalLabel label;
    private StatElementData hp;
    private StatElementData def;
    private StatElementData atk;
    private StatElementData ki;
    private StatElementData spd;

    private int scrollX;
    private int scrollWidth;

    public RaceChangeStatForm(int x, int y, int width) {
        super(width, 0);
        this.setPosition(x, y);
        this.drawBase = false;
        this.scrollX = width - 60;
        this.scrollWidth = 50;

        FormFlow scrollFlow = new FormFlow(5);

        this.label = this.addComponent(new FormLocalLabel(new LocalMessage("dragonstats", "modifierlabel"), new FontOptions(14), FormLabel.ALIGN_MID, width/2, scrollFlow.next(25)));

        this.hp = DrawStatElement(new LocalMessage("dragonstats", "hp"), scrollFlow);
        this.def = DrawStatElement(new LocalMessage("dragonstats", "def"), scrollFlow);
        this.atk = DrawStatElement(new LocalMessage("dragonstats", "atk"), scrollFlow);
        this.ki = DrawStatElement(new LocalMessage("dragonstats", "ki"), scrollFlow);
        this.spd = DrawStatElement(new LocalMessage("dragonstats", "spd"), scrollFlow);

        scrollFlow.next(5);

        this.setHeight(scrollFlow.next());

    }

    private StatElementData DrawStatElement(LocalMessage labelMessage, FormFlow scrollFlow) {
        StatElementData data = new StatElementData();
        data.label = this.addComponent(new FormLocalLabel(labelMessage, new FontOptions(12), FormLabel.ALIGN_LEFT, 20, scrollFlow.next()));
        data.value = this.addComponent(new FormHorizontalFloatScroll(scrollX, scrollFlow.next(20), scrollWidth, FormHorizontalScroll.DrawOption.value, "", 2, 1, 3));
        return data;
    }
}

File: src/main/kotlin/dragonsaga/ui/RaceChangeStatForm.kt
package dragonsaga.ui

import dragonsaga.ui.components.FormHorizontalFloatScroll
import dragonsaga.ui.models.StatElementData
import necesse.engine.localization.message.LocalMessage
import necesse.gfx.forms.Form
import necesse.gfx.forms.components.FormFlow
import necesse.gfx.forms.components.FormHorizontalScroll.DrawOption
import necesse.gfx.forms.components.FormLabel
import necesse.gfx.forms.components.localComponents.FormLocalLabel
import necesse.gfx.gameFont.FontOptions

class RaceChangeStatForm(x: Int, y: Int, width: Int) : Form(width, 0) {
    private val label: FormLocalLabel
    private val hp: StatElementData
    private val def: StatElementData
    private val atk: StatElementData
    private val ki: StatElementData
    private val spd: StatElementData
    private val scrollX: Int
    private val scrollWidth: Int

    init {
        this.setPosition(x, y)
        drawBase = false
        scrollX = width - 60
        scrollWidth = 50
        val scrollFlow = FormFlow(5)
        label = this.addComponent(FormLocalLabel(LocalMessage("dragonstats", "modifierlabel"), FontOptions(14), FormLabel.ALIGN_MID, width / 2, scrollFlow.next(25)))

        hp = DrawStatElement(LocalMessage("dragonstats", "hp"), scrollFlow)
        def = DrawStatElement(LocalMessage("dragonstats", "def"), scrollFlow)
        atk = DrawStatElement(LocalMessage("dragonstats", "atk"), scrollFlow)
        ki = DrawStatElement(LocalMessage("dragonstats", "ki"), scrollFlow)
        spd = DrawStatElement(LocalMessage("dragonstats", "spd"), scrollFlow)
        scrollFlow.next(5)
        height = scrollFlow.next()
    }

    private fun DrawStatElement(labelMessage: LocalMessage, scrollFlow: FormFlow): StatElementData =
        StatElementData(
            this.addComponent(FormLocalLabel(labelMessage, FontOptions(12), FormLabel.ALIGN_LEFT, 20, scrollFlow.next())),
            this.addComponent(FormHorizontalFloatScroll(scrollX, scrollFlow.next(20), scrollWidth, DrawOption.value, "", 2f, 1f, 3f)))
}